packages/flutter/lib/src/widgets/focus_traversal.dart

```diff
@@ -695,136 +700,317 @@ class _SortData {
-    final FocusScopeNode nearestScope = currentNode.nearestScope;
-    invalidateScopeData(nearestScope);
-    final FocusNode focusedChild = nearestScope.focusedChild;
-    if (focusedChild == null) {
-      final FocusNode firstFocus = findFirstFocus(currentNode);
-      if (firstFocus != null) {
-        _focusAndEnsureVisible(
-          firstFocus,
-          alignmentPolicy: forward
-              ? ScrollPositionAlignmentPolicy.keepVisibleAtEnd
-              : ScrollPositionAlignmentPolicy.keepVisibleAtStart,
-        );
-        return true;
+/// Base class for all sort keys for [OrderedFocusTraversalPolicy] traversal.
+///
+/// Only orders of the same type are comparable. If a set of widgets in the same
+/// [FocusScope] contains keys that are not comparable with each other, it will
+/// assert, since the ordering is undefined.
+///
+/// See also:
+///
+///  * [NumericFocusOrder], for a focus order that describes its order with a
+///    `double`.
+abstract class FocusOrder extends Diagnosticable implements Comparable<FocusOrder> {
+  /// Abstract const constructor. This constructor enables subclasses to provide
+  /// const constructors so that they can be used in const expressions.
+  const FocusOrder();
+
+  @override
+  int compareTo(FocusOrder other) {
+    assert(runtimeType == other.runtimeType,
+      "The sorting algorithm must not compare incomparable keys, since they don't know "
+      'how to order themselves relative to each other.');
+    return doCompare(other);
+  }
+
+  /// The implementation of [compareTo].
+  ///
+  /// The argument is guaranteed to be of the same type as this object and have
+  /// the same [name].
+  ///
+  /// The method should return a negative number if this object comes earlier in
+  /// the sort order than the argument; and a positive number if it comes later
+  /// in the sort order. Returning zero causes the system to use default sort
+  /// order.
+  @protected
+  int doCompare(covariant FocusOrder other);
+}
+
+/// Assigns a numerical order to a widget subtree that is using a
+/// [OrderedFocusTraversalPolicy] to set the order in which widgets should be
+///// traversed with the keyboard.
+class NumericFocusOrder extends FocusOrder {
+  /// Const constructor. This constructor enables subclasses to provide
+  /// const constructors so that they can be used in const expressions.
+  const NumericFocusOrder(this.order) : assert(order != null);
+
+  /// The numerical order to assign to the widget subtree using
+  /// [FocusTraversalOrder].
+  ///
+  /// Determines the placement of this widget in a sequence of widgets that defines
+  /// the order in which this node is traversed by the focus policy.
+  ///
+  /// Lower values will be traversed first.
+  final double order;
+
+  @override
+  int doCompare(NumericFocusOrder other) => order.compareTo(other.order);
+}
+
+/// Uses a String to assign lexical order to a widget subtree that is using a
```

Comment:
Lexical order (=alphabetical order) is language specific, see http://userguide.icu-project.org/collation.

How does this class deal with those differences? I don't think Dart's `String.compareTo` follows those language specific rules?

Reply:
Added a comment that we're using Dart string sorting.